### DISCUSSION TITLE
Add support for covariance for methods with array-like output

### DISCUSSION THREAD

**Discussion by dmitriyse** [dotnet/csharplang#2055] [Language Ideas]
### Motivation
Spans becomes popular in method signatures.
It's very usefull for performance general purpose libraries to allow covariant "array-like" output from a method.

For example modern enumerables (#2053) can be built with the covariant array-like output. 
### Proposal

```CSharp
    public struct WriteOnlySpan<T>
    {
        /// ...
    }

    public interface ITest<out T>
    {
        /// Callee output buffer
        ReadOnlySpan<T> GetData();

        /// Caller allocated buffer for output
        void GetDataToBuffer(WriteOnlySpan<T> outBuf);
    }
```

**Comment by HaloFour** [comment]
The variance of arrays in the CLR is considered one of the bigger mistakes of .NET, an unfortunate relic copied from Java.

**Comment by dmitriyse** [comment]
Variance of array is not a variance of ReadOnlySpan. Because you are unable to do the bad things.
With array you can do this:
```CSharp
            var strs = new string[10];
            var objs = (object[])strs; // Here the hell openning.
            objs[0] = new object(); // DAMAGE !!! This is DOTNET mistake inherited from JAVA!
            Console.WriteLine(strs[0]); // BOOM! Here CLR should do a lots of different additional checks to detect the DAMAGE
            // Output : 
            // Unhandled Exception: System.ArrayTypeMismatchException: 
            // Attempted to access an element as a type incompatible with the array.
```
With downcasting of ReadOnlySpan you have no such problem.
```CSharp
            var a = new string[] {"Hello!"};
            ReadOnlySpan<string> strs = a;
            ReadOnlySpan<object> objs = a;  // Should be cast but it's currently impossible in C#
            Console.WriteLine(objs[0]); // this is totally safe, no any extra check is required in CLR
```

For example "System.Runtime.CompilerServices.Unsafe.As" allows to do any wierd thing, but nobody says that is "WORST ERROR" in dotnet history.

**Comment by huoyaoyuan** [comment]
This could be summarized by "interface-like structs with variant". Some others will benifit from it as well, like ValueTask.

**Comment by ufcpp** [comment]
Related to https://github.com/dotnet/csharplang/issues/317#issuecomment-325193165, covariant structs
